lenskit/algorithms/surprise.py

```diff
@@ -0,0 +1,235 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Jun 25 12:09:10 2019
+
+@author: mohamed.cherif
+"""
+
+import logging
+import pandas as pd
+import numpy as np
+import inspect
+
+from ..matrix import ratings_surprise
+from . import Recommender
+
+_logger = logging.getLogger(__name__)
+
+
+
+class SurpriseBaseRec(Recommender):
+    """
+    Base class for surprise-backed recommenders.
+    Args:
+        delegate(implicit.RecommenderBase):
+            The delegate algorithm.
+    Attributes:
+        delegate(implicit.RecommenderBase):
+            The :py:mod:`surprise` delegate algorithm.
+        matrix_(scipy.sparse.csr_matrix):
+            The user-item rating matrix.
+        user_index_(pandas.Index):
+            The user index.
+        item_index_(pandas.Index):
+            The item index.
+    """
+
+    def __init__(self, delegate):
+        self.delegate = delegate
+
+    def fit(self, ratings):
+        dataset = ratings_surprise(ratings)
+        trainset = dataset.build_full_trainset()
+
+        _logger.info('training %s ', self.delegate)
+
+        self.delegate.fit(trainset)
+        
+        self.ratings = ratings
+        self.trainset = trainset
+        self.inner_user_index_ = trainset.all_users()
+        self.inner_item_index_ = trainset.all_items()
+
+        return self
+
+    def recommend(self, user, n=None, ratings=None):
```

Comment:
Given this implementation, it looks like this class should implement `Predictor` instead of `Recommender`, so that it can be combined with other candidate selectors. The logic in this method is a repeat of what `UnratedItemsCandidateSelector` does.